recipes/lm/sft/recipe.py

```diff
@@ -0,0 +1,227 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from __future__ import annotations
+
+from typing import final
+
+import torch
+from torch import Tensor
+from typing_extensions import override
+
+from fairseq2.composition import register_dataset_family
+from fairseq2.datasets import SequenceBatch
+from fairseq2.device import CPU, get_default_device
+from fairseq2.metrics import MetricBag
+from fairseq2.metrics.common import (
+    add_nll_loss_metric,
+    add_seq_batch_metrics,
+    update_nll_loss_metric,
+    update_seq_batch_metrics,
+)
+from fairseq2.recipe.base import RecipeContext, TrainRecipe
+from fairseq2.recipe.evaluator import EvalUnit
+from fairseq2.recipe.model import RecipeModel
+from fairseq2.recipe.trainer import Trainer, TrainUnit
+from fairseq2.runtime.dependency import DependencyContainer
+from fairseq2.utils.rng import RngBag
+
+from .dataset import (
+    LM_SFT_PADDED_DATASET,
+    DataReadOptions,
+    LMSFTDataset,
+    LMSFTDatasetConfig,
+    open_lm_sft_dataset,
+)
+from .default_config import LMSFTConfig
+from .utils import Batching, LengthBatching, StaticBatching
+
+
+@final
+class LMSFTRecipe(TrainRecipe):
+    @override
+    def register(self, container: DependencyContainer) -> None:
+        register_dataset_family(
+            container,
+            LM_SFT_PADDED_DATASET,
+            LMSFTDataset,
+            LMSFTDatasetConfig,
+            opener=open_lm_sft_dataset,
+        )
+
+    @override
+    def create_trainer(self, context: RecipeContext) -> Trainer:
+        config = context.config.as_(LMSFTConfig)
+
+        unit = LMSFTUnit(context.model)
+
+        dataset = context.default_dataset.as_(LMSFTDataset)
+
+        seed = config.common.seed
+        device = get_default_device()
+
+        if device.type == "cuda":
+            torch.cuda.set_device(device)
+
+        rng_bag = RngBag.from_device_defaults(CPU, device)
+
+        rng_bag.manual_seed(seed)
```

Comment:
This is also not needed. fairseq2 calls `manual_seed(config.common.seed)` before calling `create_trainer`.